### ISSUE TITLE
Export Multiple Instances of Meshes in Correct Location?

### ISSUE THREAD

**Issue by SceneForgeStudio** [atteneder/glTFast#364]
**Is your feature request related to a problem? Please describe.**
In my app, users can instantiate prefab 3D models and place them wherever they like.
The problem is when exporting the scene (an array of desired gameobjects),  despite the array containing multiple of the same mesh in different positions around the scene, in the export there's only one instance of that mesh, all bunched in the same location.  With a few odd exceptions (see the window and door meshes in the images below).

**Describe the solution you'd like**
I understand the value of the current implementation, but for level builders, it doesn't quite work.  
Maybe another bool flag in the export settings to respect multiple instances of the same model in different locations?

**Describe alternatives you've considered**
I couldn't figure out a good way to do this currently.

**Additional context**
Here's a pic of the scene in Unity in comparison to what the export looks like
![image](https://user-images.githubusercontent.com/40009793/157498149-2b773612-1443-4961-8946-63337b53fa83.png)

![image](https://user-images.githubusercontent.com/40009793/157524888-1ec6382e-7fe5-4cda-a1f1-c04b21489f82.png)

**Comment by kyapp69** [comment]
Question. Do your harps share the same mesh? As long as they share the same mesh, there's scope to export one mesh and use the same accesors for for different nodes. That would make sense and also make the file size smaller.

**Comment by atteneder** [comment]
> As long as they share the same mesh, there's scope to export one mesh and use the same accesors for for different nodes. That would make sense and also make the file size smaller.

That's true and it should already work like that.

@SceneForgeStudio We'll have to figure out why glTFast does not detect the shared mesh of the three harps. Can you provide a minimal test scene that reproduces this behavior?

**Comment by atteneder** [comment]
Changed this to a bug.

Note: With identical materials (and a greater number of instances) it would be cool to use the instancing extension instead (or have it as an option), but that's another story.

**Comment by SceneForgeStudio** [comment]
@kyapp69 Yes, they are exact duplicates just positioned in different locations.  
The one caveat being that the object is imported at runtime using Trilib.  That being said, I don't believe that simply instantiating the same mesh would break it.

@atteneder I will do some more testing in isolated scenes and update this issue with an upload.  Thanks!

**Comment by kyapp69** [comment]
I have been doing some test.. @SceneForgeStudio can you verify that it's still happening? Are there any skinned meshes? If the meshes are skinned it can happen. I tested with just pure non skinned meshes and they seemed to export ok to their respective places.

**Comment by SceneForgeStudio** [comment]
@kyapp69 So sorry I didn't get a chance to upload a reproducible sample, life got in the way and I was away from my desk.
None of the meshes are skinned, just regular meshes.  How was your test done?  Using prefabs?  Like I said, the models are imported at runtime using Trilib, then I'm duplicating (Instantiating) the models.
What steps are you taking that arent resulting in this happening?

**Comment by SceneForgeStudio** [comment]
> Turns out a faulty GetHashCode implementation led to duplication of glTF meshes. It's fixed in `main` and will be released soonish.
> 
> Again, thanks for reporting!

Hey, sorry for not being active and uploading any samples, but I'm back to working on exporting, and am having the same issue using the latest version (4.8.2).

You can see the examples below:

The scene is made up of the following meshes:
- 1 Floor Plane
- 4 Walls (one has a door frame cut out
- 1 Door Frame
- 1 Door
- 2 Door Handles
But you can see there are extra door meshes and a frame.
![image](https://user-images.githubusercontent.com/40009793/177059499-48dcbfcc-b7db-4145-aea7-e1aa0ac8cb19.png)

In another example, I have four duplicate objects next to each other added to the export scene, but the resulting file has all 4 meshes in the same place.
It looks like a single object, but viewing the file in a more advanced GLB viewer online shows that all 4 are present.

![image](https://user-images.githubusercontent.com/40009793/177059632-ef6e8447-5df3-4d20-892e-bf35c032f221.png)

Additionally, in the example below, you can see I have the desired meshes inside the room object, but the result has them super far away for some reason.

![image](https://user-images.githubusercontent.com/40009793/177059712-40a71d49-3807-4735-953d-1401a1479669.png)

I'm using the following method to export a desired list of GameObjects (and hiding children that aren't wanted).
For standalone objects I make sure to add all desired child meshes to the `object` list.  Same goes for the room object (all doors, etc).
Is there something I'm doing incorrectly?

```
 		public static async void ExportGLTF(string fileName, string extention, List<GameObject> objects, bool exportChildren = true)
        {
            List<bool> activeStates = new List<bool>();
            List<List<MeshRenderer>> allChildrenLists = new List<List<MeshRenderer>>();
            int meshIndex = 0;

            for (int i = 0; i < objects.Count; i++)
            {
                GameObject obj = objects[i];
                var mrs = obj.GetComponentsInChildren<MeshRenderer>(true);
                allChildrenLists.Add(mrs.ToList());
            }

            var exportSettings = new ExportSettings
            {
                format = extention.ToLower() == ".gltf" ? GltfFormat.Json : GltfFormat.Binary,
                fileConflictResolution = FileConflictResolution.Overwrite,
                imageDestination = ImageDestination.MainBuffer
            };
            var goExport = new GameObjectExportSettings
            {
                onlyActiveInHierarchy = true
            };

            if (!exportChildren)
                ToggleChildren(false);

            var export = new GameObjectExport(exportSettings, goExport);
            export.AddScene(objects.ToArray(), "My Scene");
            string p = fileName + extention;
            
            bool success = await export.SaveToFileAndDispose(p);
			// here I call success/fail events if needed

            if (!exportChildren)
                ToggleChildren(true);

            void ToggleChildren(bool on)
            {
                if (on)
                    meshIndex = 0;

                for (int i = 0; i < allChildrenLists.Count; i++)
                {
                    for (int j = 0; j < allChildrenLists[i].Count; j++)
                    {
                        var r = allChildrenLists[i][j].gameObject;
                        if (objects.Contains(r))
                            continue;

                        if (on == false) // meaning it's the start of an export
                        {
                            //allMeshes.Add(r);
                            activeStates.Add(r.activeSelf);
                            r.gameObject.SetActive(false);
                        }
                        else
                        {
                            bool activeState = activeStates[meshIndex];

                            if (activeState)
                                r.SetActive(on);

                            meshIndex++;
                        }
                    }
                }
            }
```
Thanks!

**Comment by atteneder** [comment]
@SceneForgeStudio 

Thanks for reporting.

The current implementation of `GameObjectExport.AddScene` adds all provided GameObjects assuming they are root Transforms and using their local transform. I assume in your source scene, they are not root objects, nested somewhere deeper in the hierarchy and thus have additional transformation on them that got lost.

I agree that this might be confusing / not expected. What other behavior would you expect/make sense? My guess:

- Use world transforms
- Add all upstream parent GameObjects without their components nor other children
- Make users add the (common) root object and provide advanced include/exclude filtering.
- Don't change anything, but properly rename parameters and document

So far I don't really like any option but the last, but I'm open for feedback and further suggestions.

@hybridherbst I checked UnityGLTF and saw it's behaving the same way (`GLTFSceneExporter.ExportScene`). Any idea/opinion?

Thanks

**Comment by SceneForgeStudio** [comment]
> @SceneForgeStudio
> 
> Thanks for reporting.
> 
> The current implementation of `GameObjectExport.AddScene` adds all provided GameObjects assuming they are root Transforms and using their local transform. I assume in your source scene, they are not root objects, nested somewhere deeper in the hierarchy and thus have additional transformation on them that got lost.
> 
> I agree that this might be confusing / not expected. What other behavior would you expect/make sense? My guess:
> 
> * Use world transforms
> * Add all upstream parent GameObjects without their components nor other children
> * Make users add the (common) root object and provide advanced include/exclude filtering.
> * Don't change anything, but properly rename parameters and document
> 
> So far I don't really like any option but the last, but I'm open for feedback and further suggestions.
>  
> Thanks

Ah, interesting, thank you for the info.
You are correct, I keep all scene objects under a "SceneParent" object for neatness' sake, and then reference the objects I want to import within that parent.  Then, within those objects are children that I want to export as well (with some that I don't)

I suppose I can unparent and reparent the objects upon exporting in the same way I am setting them active/inactive.  Would that work in theory?  Not a super clean solution though.

If I have a hierarchy like so:
```
-- SceneParent (Root)
--- >  Object Parent
-------> Export
-------> Export
-------> Don't Export
------------> Empty Child
------------------> Export
------------------>  Don't Export
--- > Object Parent
// And so on
```

How would you best recommend I export it accordingly?

Thanks so much!

**Comment by atteneder** [comment]
> I suppose I can unparent and reparent the objects upon exporting in the same way I am setting them active/inactive. Would that work in theory? Not a super clean solution though.

Yes, that would work. For a lack of alternatives, I'd recommend it for now.
Such hierarchy changes can have a performance hit. That isn't a big deal if this isn't done often.

> If I have a hierarchy like so:
> 
> ```
> -- SceneParent (Root)
> --- >  Object Parent
> -------> Export
> -------> Export
> -------> Don't Export
> ------------> Empty Child
> ------------------> Export
> ------------------>  Don't Export
> --- > Object Parent
> // And so on
> ```
> 
> How would you best recommend I export it accordingly?

I'd use the export setting `onlyActiveInHierarchy`, set all objects you don't want included inactive (temporarily), export and restore the prior active state on all GameObjects.

Thanks for showing that use-case. It'll help making the next API version.

**Comment by SceneForgeStudio** [comment]
> > I suppose I can unparent and reparent the objects upon exporting in the same way I am setting them active/inactive. Would that work in theory? Not a super clean solution though.
> 
> Yes, that would work. For a lack of alternatives, I'd recommend it for now. Such hierarchy changes can have a performance hit. That isn't a big deal if this isn't done often.
> 
> > If I have a hierarchy like so:
> > ```
> > -- SceneParent (Root)
> > --- >  Object Parent
> > -------> Export
> > -------> Export
> > -------> Don't Export
> > ------------> Empty Child
> > ------------------> Export
> > ------------------>  Don't Export
> > --- > Object Parent
> > // And so on
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How would you best recommend I export it accordingly?
> 
> I'd use the export setting `onlyActiveInHierarchy`, set all objects you don't want included inactive (temporarily), export and restore the prior active state on all GameObjects.
> 
> Thanks for showing that use-case. It'll help making the next API version.

Thank you for the suggestion and info.
Would doing that fix the strange bug with the extra door meshes?

**Comment by atteneder** [comment]
> Thank you for the suggestion and info. Would doing that fix the strange bug with the extra door meshes?

Yes, that should work.

**Comment by SceneForgeStudio** [comment]
Amazing, after much trial and error, this works consistently every time.
Here's my full method for anyone else searching for a good way to do this:

```
 public static async void ExportGLTF(string fileName, string extention, List<GameObject> objects, Action<string> onComplete = null, Action onFail = null)
        {
            List<bool> activeStates = new List<bool>();
            List<List<MeshRenderer>> allChildrenLists = new List<List<MeshRenderer>>();
            List<Transform> originalParents = new List<Transform>();

            int meshIndex = 0;

            for (int i = 0; i < objects.Count; i++)
            {
                GameObject obj = objects[i];
                var mrs = obj.GetComponentsInChildren<MeshRenderer>(true);
                allChildrenLists.Add(mrs.ToList());
                originalParents.Add(obj.transform.parent);
                obj.transform.parent = null;
            }

            var exportSettings = new ExportSettings
            {
                format = extention.ToLower() == ".gltf" ? GltfFormat.Json : GltfFormat.Binary,
                fileConflictResolution = FileConflictResolution.Overwrite,
                imageDestination = ImageDestination.MainBuffer
            };
            var goExport = new GameObjectExportSettings
            {
                onlyActiveInHierarchy = true
            };

            SetupMeshes(false);

            var export = new GameObjectExport(exportSettings, goExport);
            export.AddScene(objects.ToArray(), "My Scene");
            string p = fileName + extention;
            
            bool success = await export.SaveToFileAndDispose(p);

            if (success)
                onComplete?.Invoke(fileName);
            else
                onFail?.Invoke();

            SetupMeshes(true);

            void SetupMeshes(bool exportComplete)
            {
                if (exportComplete)
                    meshIndex = 0;

                for (int i = 0; i < allChildrenLists.Count; i++)
                {
                    for (int j = 0; j < allChildrenLists[i].Count; j++)
                    {
                        var r = allChildrenLists[i][j].gameObject;
                        if (objects.Contains(r))
                            continue;

                        if (exportComplete == false) // meaning it's the start of an export
                        {
                            //allMeshes.Add(r);
                            activeStates.Add(r.activeSelf);
                            r.gameObject.SetActive(false);
                        }
                        else
                        {
                            bool activeState = activeStates[meshIndex];

                            if (activeState)
                                r.SetActive(activeState);

                            meshIndex++;
                        }
                    }
                }

                if (exportComplete)
                {
                    for (int i = 0; i < objects.Count; i++)
                    {
                        var parents = originalParents;
                        objects[i].transform.parent = parents[i];
                    }
                }               
            }
        }
```
